ledger/participant-integration-api/src/main/scala/platform/store/ActiveLedgerStateManager.scala

```diff
@@ -69,146 +105,184 @@ private[platform] class ActiveLedgerStateManager[ALS <: ActiveLedgerState[ALS]](
-                )
-              case nlkup: N.NodeLookupByKey[ContractId] =>
-                // Check that the stored lookup result matches the current result
-                val key = nlkup.key.key.ensureNoCid.fold(
-                  coid =>
-                    throw new IllegalStateException(s"Contract ID $coid found in contract key"),
-                  identity,
+                val newState = acc.copy(
+                  errs = contractCheck(als, ne.targetCoid).fold(acc.errs)(acc.errs + _),
+                  currentState = acc.currentState.copy(
+                    als = Some(if (ne.consuming) {
+                      als.removeContract(ne.targetCoid)
+                    } else {
+                      als
+                    }),
+                    archivedIds =
+                      if (ne.consuming) acc.currentState.archivedIds + ne.targetCoid
+                      else acc.currentState.archivedIds,
+                  ),
+                  parties = acc.parties.union(nodeParties),
                 )
-                val gk = GlobalKey(nlkup.templateId, key)
-                val nodeParties = nlkup.key.maintainers
-
-                if (actAs.nonEmpty) {
-                  // If the submitter is known, look up the contract
-                  // Submitters being know means the transaction was submitted on this participant.
-                  val currentResult = acc.lookupContractByKey(gk)
-                  if (currentResult == nlkup.result) {
-                    ats.copy(
-                      parties = parties.union(nodeParties)
-                    )
-                  } else {
-                    ats.copy(
-                      errs = errs + Inconsistent(
-                        s"Contract key lookup with different results: expected [${nlkup.result}], actual [${currentResult}]"
-                      )
-                    )
-                  }
-                } else {
-                  // Otherwise, trust that the lookup was valid.
-                  // The submitter being unknown means the transaction was submitted on a different participant,
-                  // and (A) this participant may not know the authoritative answer to whether the key exists and
-                  // (B) this code is called from a Indexer and not from the sandbox ledger.
-                  ats.copy(
-                    parties = parties.union(nodeParties)
-                  )
-                }
+                (newState, true)
             }
-        }
-
-    val divulgedContractIds = divulgence -- st.archivedIds
+          },
+          rollbackBegin = (acc, _, _) => {
+            (acc.copy(rollbackStates = acc.currentState +: acc.rollbackStates), true)
+          },
+          leaf = (acc, nodeId, node) => handleLeaf(acc, nodeId, node),
+          exerciseEnd = (acc, _, _) => acc,
+          rollbackEnd = (acc, _, _) => {
+            val (beforeRollback, rest) = acc.rollbackStates match {
```

Comment:
Prefer vanilla matching with `Nil` and `::` to avoid disabling exhaustive pattern match checking.
```
              case Nil => sys.error("IMPOSSIBLE: Rollback end but rollback stack is empty")
              case head :: tail => (head,tail)
```
Currently the `Seq()` case can be removed and the compiler won't complain. Scala, eh? :man_shrugging:

Reply:
b723f0518735980846e2b6d3749402541660bca8